File: robert-server-ws-rest/src/test/java/fr/gouv/stopc/robertserver/ws/test/StatisticsManager.java
package fr.gouv.stopc.robertserver.ws.test;

import io.restassured.response.ValidatableResponse;
import org.assertj.core.api.AbstractIntegerAssert;
import org.jetbrains.annotations.NotNull;
import org.springframework.test.context.TestContext;
import org.springframework.test.context.TestExecutionListener;

import static io.restassured.RestAssured.given;
import static org.assertj.core.api.Assertions.assertThat;

public class StatisticsManager implements TestExecutionListener {

    private static ValidatableResponse todaySnapshot;

    @Override
    public void beforeTestMethod(@NotNull TestContext testContext) {
        todaySnapshot = getTodayStatistics();
    }

    public static AbstractIntegerAssert<?> assertThatTodayStatistic(String statisticName) {
        final int originalValue = todaySnapshot.extract().path(statisticName);
        final int currentValue = getTodayStatistics().extract().path(statisticName);
        return assertThat(currentValue - originalValue)
                .as("'%s' statistic value increment", statisticName);
    }

    private static ValidatableResponse getTodayStatistics() {
        return given()
                .get("/internal/api/v2/kpis")
                .then();
    }
}

File: robert-server-ws-rest/src/test/kotlin/fr/gouv/stopc/robertserver/ws/test/StatisticsManager.kt
package fr.gouv.stopc.robertserver.ws.test

import fr.gouv.stopc.robertserver.ws.repository.model.KpiName
import fr.gouv.stopc.robertserver.ws.repository.model.KpiName.*
import io.restassured.RestAssured.given
import io.restassured.response.ValidatableResponse
import org.assertj.core.api.AbstractIntegerAssert
import org.assertj.core.api.Assertions.assertThat
import org.springframework.test.context.TestContext
import org.springframework.test.context.TestExecutionListener

class StatisticsManager : TestExecutionListener {

    companion object {

        private lateinit var todaySnapshot: ValidatableResponse

        fun assertThatTodayStatistic(statisticName: KpiName?): AbstractIntegerAssert<*> {
            val originalValue = todaySnapshot.extract().path<Int>(statisticName?.key)
            val currentValue = getStatistics().extract().path<Int>(statisticName?.key)
            return assertThat(currentValue - originalValue)
                .`as`("'${statisticName?.key}' statistic value increment")
        }

        fun assertThatTodayStatisticsDidntIncrement(kpis: List<KpiName>) {
            kpis.map { kpi -> assertThatTodayStatistic(kpi).isEqualTo(0) }
        }

        private fun getStatistics() = given()
            .get("/internal/api/v2/kpis")
            .then()
    }

    override fun beforeTestMethod(testContext: TestContext) {
        todaySnapshot = getStatistics()
    }
}
